### DISCUSSION TITLE
Example of inheriting from "bundled" plugins to create custom plugins

### DISCUSSION THREAD

**Discussion by raidenmiro** [kubb-labs/kubb#1074] [General]
It's a very cool too 🔥. Enjoyed working with him.

There's a question that I haven't fully realized yet. I want to take **pluginClient** from **@kubb/swagger-client** as a basis, and for my tasks, provide a wrapper that will extend pluginClient, with custom template and parameters. Are there any examples, ideas of how to implement this correctly? I would be very grateful  ☺️

**Comment by stijnvanhulle** [comment]
Maybe this can help in understanding Kubb plugins(just some basics): https://www.kubb.dev/plugins/development/system#setup. You can call a plugin from another plugin(that is how we do it with some internal plugins) but there is no documentation. Templates could maybe also help out for you: https://www.kubb.dev/plugins/swagger-client/#templates.

**Reply by raidenmiro** [reply]
Thanks, I'll look into the internal plugins.

**Reply by raidenmiro** [reply]
@stijnvanhulle, Hi! I found the `transform` method in the lifecycle documentation https://www.kubb.dev/reference/pluginManager/lifecycle#transform. I'm trying to extend the typescript plugin and somehow transform the source code after generation (ts compiler api). Everything works as it should, as long as I don't go into other plugins like swagger-client

```ts
const bundledTsPlugin = (options: EnhancedTsOptions) => {
  const plugin = pluginTs(options)
  
  plugin.transform = function(_path, code) {
    // do transform source code
    return newSourceCode
  }
  
  return plugin
}
```

Transform is working correctly

```ts
export default definePlugin({
  plugins: [pluginOs(), bundledTsPlugin()]
})
```

The transformation is overridden by another plugin

```ts
export default definePlugin({
  plugins: [pluginOs(), bundledTsPlugin(), pluginClient()]
})
```

any ideas on how I can fight this? Maybe there is some api more correct for such a case? 🙏

**Reply by stijnvanhulle** [reply]
Transform was added but has not been used yet in Kubb itself so normally the last plugin has the power of the transform so maybe move your plugin to the last position? Not sure what you are trying to do and if Kubb can support your use case(the plugin system is only used in Kubb internal plugins)

```
export default definePlugin({
  plugins: [pluginOs(), pluginClient(), bundledTsPlugin()]
})
```

**Reply by raidenmiro** [reply]
Thanks, maybe using `transform` is not a good solution for me. I use it to convert source code using the typescript compiler api. 

Example:

```ts
const printer = ts.createPrinter()

const bundledTsPlugin = (options: EnhancedTsOptions) => {
  const plugin = pluginTs(options)
  
  plugin.transform = function(_path, code) {
    const sourceFile = ts.createSourceFile('temp.ts', code, ts.ScriptTarget.Latest);
    
    // convert all enum members to pascal case convention
    const transformationResult = ts.transform(sourceFile, [pascalizeEnumMembersName]);
    //. Here we can modify the generation results in some way

    return printer.printFile(transformationResult[0]);
  }
  
  return plugin
}
```

Can we consider expanding the transform function in pluginTs options, like: https://openapi-ts.dev/node#transform-posttransform? 
It will often be useful to override the schema in some way 🤔.
